### PR TITLE
365 default imputation links

### PR DESCRIPTION
**Pull request by Jday7879**
# Summary

Implemented method for calculating default imputation link values based on specifications

# Checklists

This pull request meets the following requirements:

- [x] installable with all dependencies recorded
- [x] runs without error
- [x] follows PEP8 and project specific conventions
- [x] appropriate use of comments, for example no descriptive comments
- [x] functions documented using Numpy style docstings
- [x] assumptions and decisions log considered and updated if appropriate
- [x] unit tests have been updated to cover essential functionality for a reasonable range of inputs and conditions
- [x] other forms of testing such as end-to-end and user-interface testing have been considered and updated as required
- [x] tests suite passes (locally as a minimum)
- [x] peer reviewed with review recorded

If you feel some of these conditions do not apply for this pull request, please
add a comment to explain why.

### CODE REVIEW

File: mbs_results/calculate_imputation_link.py [hunk 1]
```diff
@@ -44,6 +46,15 @@ def calculate_imputation_link(
 
     df_intermediate = df.copy()
 
+    # Need this for edge case when denominator is zero
+    # But is from a valid zero return
+    df_intermediate["true_zeros_column"] = df_intermediate[predictive_variable] == 0
```
Comment (rowanhemsi) [inline mbs_results/calculate_imputation_link.py:51]: I think I understand this syntax but it could be more explicit

File: mbs_results/calculate_imputation_link.py [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import numpy as np
 import pandas as pd
 
+from mbs_results.flag_and_count_matched_pairs import count_matches
+
 
 def calculate_imputation_link(
```
Comment (rowanhemsi) [inline mbs_results/calculate_imputation_link.py:7]: This is now a pretty long function, consider abstracting some of the new functionality into a lower level function

File: tests/test_ratio_of_means.py [hunk 3]
```diff
@@ -99,10 +94,18 @@ def test_ratio_of_means(self, base_file_name):
...
             sum, axis=1
         )
+        actual_output = actual_output.rename(
+            columns={
+                "default_link_b_match": "default_backward",
+                "default_link_f_match": "default_forward",
+                "default_link_flag_construction_matches": "default_construction",
+                "flag_construction_matches_pair_count": "flag_match_pair_count",
+            }
+        )
```
Comment (rowanhemsi) [inline tests/test_ratio_of_means.py:108]: Why do these need to be renamed?

**Review by rowanhemsi — COMMENTED** [review]
Left a couple of suggestions but this seems to cover the edge cases we are worried about as far as I understand them :)
The spec says that the matched pairs should be either missing or zero to indicate what type of default link case has happened - is this covered in the tests?

File: mbs_results/calculate_imputation_link.py (hunk 2 above)
Comment (Jday7879) [reply mbs_results/calculate_imputation_link.py:7]: Extracted some of the functionality into a new function

File: tests/test_ratio_of_means.py (hunk 3 above)
Comment (Jday7879) [reply tests/test_ratio_of_means.py:108]: The unit test for ratio of means are taken from the Pyspark versions which use different column names. It might be worth to bulk change these at some point, but there are about 70 files which would need to be updated. Happy to add this onto the tech backlog

**Comment by Jday7879** [comment]
> Left a couple of suggestions but this seems to cover the edge cases we are worried about as far as I understand them :) The spec says that the matched pairs should be either missing or zero to indicate what type of default link case has happened - is this covered in the tests?

Unit test data is taken from the Pyspark method and have been built to account for zeros and nulls present in the data (From my understanding)

Updated the code in line with previous comments

File: tests/test_ratio_of_means.py (hunk 3 above)
Comment (rowanhemsi) [reply tests/test_ratio_of_means.py:108]: That makes sense. I agree that it's worth changing them but it's definitely not the top priority

**Review by rowanhemsi — APPROVED** [review]
Thanks for responding to my comments, happy to approve :)
